### ISSUE TITLE
DrawRef throws System.ArgumentException   

### ISSUE THREAD

**Issue by jbialobr** [gitextensions/gitextensions#5276]
**Do you want to request a *feature* or report a *bug*?**

:bug:

**What is the current behavior?**

```
System.ArgumentException
  HResult=0x80070057
  Message=Parametr jest nieprawidłowy.
  Source=System.Drawing
  Ślad stosu:
   at System.Drawing.Graphics.get_TextRenderingHint()
   at System.Windows.Forms.Internal.WindowsFont.WindowsFontQualityFromTextRenderingHint(Graphics g)
   at System.Windows.Forms.TextRenderer.MeasureText(IDeviceContext dc, String text, Font font, Size proposedSize, TextFormatFlags flags)
   at GitUI.RevisionGridRefRenderer.DrawRef(Boolean isRowSelected, Font font, Int32& offset, String name, Color headColor, RefArrowType arrowType, Rectangle& bounds, Graphics graphics, Boolean dashedLine, Boolean fill) in D:\projekty\GitExt\GitUI\UserControls\RevisionGrid\RevisionGridRefRenderer.cs:line 26
   at GitUI.UserControls.RevisionGrid.Columns.MessageColumnProvider.OnCellPainting(DataGridViewCellPaintingEventArgs e, GitRevision revision, Int32 rowHeight, ValueTuple`4& style) in D:\projekty\GitExt\GitUI\UserControls\RevisionGrid\Columns\MessageColumnProvider.cs:line 183
   at GitUI.UserControls.RevisionGrid.RevisionDataGridView.OnCellPainting(Object sender, DataGridViewCellPaintingEventArgs e) in D:\projekty\GitExt\GitUI\UserControls\RevisionGrid\RevisionDataGridView.cs:line 314
   at System.Windows.Forms.DataGridView.OnCellPainting(DataGridViewCellPaintingEventArgs e)
   at System.Windows.Forms.DataGridViewCell.PaintWork(Graphics graphics, Rectangle clipBounds, Rectangle cellBounds, Int32 rowIndex, DataGridViewElementStates cellState, DataGridViewCellStyle cellStyle, DataGridViewAdvancedBorderStyle advancedBorderStyle, DataGridViewPaintParts paintParts)
   at System.Windows.Forms.DataGridViewRow.PaintCells(Graphics graphics, Rectangle clipBounds, Rectangle rowBounds, Int32 rowIndex, DataGridViewElementStates rowState, Boolean isFirstDisplayedRow, Boolean isLastVisibleRow, DataGridViewPaintParts paintParts)
   at System.Windows.Forms.DataGridViewRow.Paint(Graphics graphics, Rectangle clipBounds, Rectangle rowBounds, Int32 rowIndex, DataGridViewElementStates rowState, Boolean isFirstDisplayedRow, Boolean isLastVisibleRow)
   at System.Windows.Forms.DataGridView.PaintRows(Graphics g, Rectangle boundingRect, Rectangle clipRect, Boolean singleHorizontalBorderAdded)
   at System.Windows.Forms.DataGridView.PaintGrid(Graphics g, Rectangle gridBounds, Rectangle clipRect, Boolean singleVerticalBorderAdded, Boolean singleHorizontalBorderAdded)
   at System.Windows.Forms.DataGridView.OnPaint(PaintEventArgs e)
```

![gitext](https://user-images.githubusercontent.com/1003909/43580265-196ec190-9655-11e8-903e-b148b60c75a1.png)

**Environment you encounter the issue:**

- GitExtensions version: fe59613e642a67636f325174ab214a723a82223e

**Comment by RussKie** [comment]
Under what circumstances have you observed this?

**Comment by drewnoakes** [comment]
Decompiled WinForms code:

```c#
    public TextRenderingHint TextRenderingHint
    {
      get
      {
        TextRenderingHint textRenderingHint1 = TextRenderingHint.SystemDefault;
        int textRenderingHint2 = SafeNativeMethods.Gdip.GdipGetTextRenderingHint(new HandleRef((object) this, this.NativeGraphics), out textRenderingHint1);
        if (textRenderingHint2 != 0)
          throw SafeNativeMethods.Gdip.StatusException(textRenderingHint2);
        return textRenderingHint1;
      }
```

The non-zero return code from `GdipGetTextRenderingHint` is turned into an exception in `StatusException`. My Polski's not so hot, but _"Parametr jest nieprawidłowy"_ looks like a generic parameter exception, so it's hard to tell which of these it might be (probably `case 2`):

```c#
      internal static Exception StatusException(int status)
      {
        switch (status)
        {
          case 1:
            return (Exception) new ExternalException(SR.GetString("GdiplusGenericError"), -2147467259);
          case 2:
            return (Exception) new ArgumentException(SR.GetString("GdiplusInvalidParameter"));
          case 3:
            return (Exception) new OutOfMemoryException(SR.GetString("GdiplusOutOfMemory"));
          case 4:
            return (Exception) new InvalidOperationException(SR.GetString("GdiplusObjectBusy"));
          case 5:
            return (Exception) new OutOfMemoryException(SR.GetString("GdiplusInsufficientBuffer"));
          case 6:
            return (Exception) new NotImplementedException(SR.GetString("GdiplusNotImplemented"));
          case 7:
            return (Exception) new ExternalException(SR.GetString("GdiplusGenericError"), -2147467259);
          case 8:
            return (Exception) new InvalidOperationException(SR.GetString("GdiplusWrongState"));
          case 9:
            return (Exception) new ExternalException(SR.GetString("GdiplusAborted"), -2147467260);
          case 10:
            return (Exception) new FileNotFoundException(SR.GetString("GdiplusFileNotFound"));
          case 11:
            return (Exception) new OverflowException(SR.GetString("GdiplusOverflow"));
          case 12:
            return (Exception) new ExternalException(SR.GetString("GdiplusAccessDenied"), -2147024891);
          case 13:
            return (Exception) new ArgumentException(SR.GetString("GdiplusUnknownImageFormat"));
          case 14:
            return (Exception) new ArgumentException(SR.GetString("GdiplusFontFamilyNotFound", new object[1]
            {
              (object) "?"
            }));
          case 15:
            return (Exception) new ArgumentException(SR.GetString("GdiplusFontStyleNotFound", (object) "?", (object) "?"));
          case 16:
            return (Exception) new ArgumentException(SR.GetString("GdiplusNotTrueTypeFont_NoName"));
          case 17:
            return (Exception) new ExternalException(SR.GetString("GdiplusUnsupportedGdiplusVersion"), -2147467259);
          case 18:
            return (Exception) new ExternalException(SR.GetString("GdiplusNotInitialized"), -2147467259);
          case 19:
            return (Exception) new ArgumentException(SR.GetString("GdiplusPropertyNotFoundError"));
          case 20:
            return (Exception) new ArgumentException(SR.GetString("GdiplusPropertyNotSupportedError"));
          default:
            return (Exception) new ExternalException(SR.GetString("GdiplusUnknown"), -2147418113);
        }
      }
```

It's hard to tell what the problem here is, or if it's something specific to `DrawRef`.

The `Graphics.TextRenderingHint` takes no arguments, so for it to throw presumably it must be in a bad state. That instance comes from `DataGridViewCellPaintingEventArgs`, which is provided to us. There's nothing asynchronous going on in that call stack that could get things out of order for us. I wonder if this is an internal exception that it recovers from on its own, given that continuing didn't give the usual WinForms red X, or crash the app.

In short, I see nothing here that we can deal with easily, and I like to think you only saw this exception because you were running the debugger, and that user's wouldn't be affected by this.

**Comment by drewnoakes** [comment]
Going to close this for now as can't find any path forward. If someone is able to reproduce it in the debugger, glean more information and determine there's something we can do about it, let's re-open then.
